include/sparrow/arrow_array_schema_proxy.hpp

```diff
@@ -95,99 +139,176 @@ namespace sparrow
+        const auto buffer_count = static_cast<size_t>(m_array_ref.n_buffers);
         m_buffers.reserve(buffer_count);
-        const data_type data_type = format_to_data_type(schema().format);
+        const enum data_type data_type = format_to_data_type(m_schema_ref.format);
         const std::vector<buffer_type> buffers_type = get_buffer_types_from_data_type(data_type);
         for (std::size_t i = 0; i < buffer_count; ++i)
         {
             const auto buffer_type = buffers_type[i];
-            auto buffer = array().buffers[i];
-            const std::size_t buffer_size = compute_buffer_size(buffer_type, length(), data_type);
+            auto buffer = m_array_ref.buffers[i];
+            const std::size_t buffer_size = compute_buffer_size(buffer_type, length(), offset(), data_type);
             auto* ptr = static_cast<uint8_t*>(const_cast<void*>(buffer));
             m_buffers.emplace_back(ptr, buffer_size);
         }
     }
 
-    inline
-    void arrow_proxy::validate_array_and_schema() const
+    bool arrow_proxy::array_created_with_sparrow() const
     {
-        SPARROW_ASSERT_TRUE(this->array().release != nullptr);
-        SPARROW_ASSERT_TRUE(this->schema().release != nullptr);
+        return m_array_ref.release == &sparrow::release_arrow_array;
+    }
 
-        SPARROW_ASSERT_TRUE(this->array().n_children == this->schema().n_children);
-        SPARROW_ASSERT_TRUE((this->array().dictionary == nullptr) == (this->schema().dictionary == nullptr));
+    bool arrow_proxy::schema_created_with_sparrow() const
+    {
+        return m_schema_ref.release == &sparrow::release_arrow_schema;
+    }
+
+    void arrow_proxy::validate_array_and_schema() const
+    {
+        SPARROW_ASSERT_TRUE(m_array_ref.release != nullptr);
+        SPARROW_ASSERT_TRUE(m_schema_ref.release != nullptr);
+        SPARROW_ASSERT_TRUE(m_array_ref.n_children == m_schema_ref.n_children);
+        SPARROW_ASSERT_TRUE((m_array_ref.dictionary == nullptr) == (m_schema_ref.dictionary == nullptr));
 
-        const data_type data_type = format_to_data_type(this->schema().format);
-        if (!validate_format_with_arrow_array(data_type, array()))
+        const enum data_type data_type = format_to_data_type(m_schema_ref.format);
+        if (!validate_format_with_arrow_array(data_type, m_array_ref))
         {
-            // TODO: Replace with a more specific exception
-            throw std::runtime_error("Invalid ArrowArray format");
+            throw arrow_proxy_exception("Invalid ArrowArray format");
         }
     }
 
-    inline
-    arrow_proxy::arrow_proxy(ArrowArray&& array, ArrowSchema&& schema)
+    inline arrow_proxy::arrow_proxy(ArrowArray&& array, ArrowSchema&& schema)
         : m_array(std::move(array))
+        , m_array_ref(std::get<1>(m_array))
         , m_schema(std::move(schema))
+        , m_schema_ref(std::get<1>(m_schema))
+        , m_data_ownership({.schema = ownership::owning, .array = ownership::owning})
```

Comment:
```suggestion
        , m_data_ownership(owns_arrow_data)
```
See these constants designed specifically for that: 
- [`owns_arrow_data`](https://github.com/man-group/sparrow/blob/11ef0540b3227b79cf66fb0f331f493ff488b6de/include/sparrow/c_interface.hpp#L109C27-L109C43)
- [`doesnt_own_arrow_data`](https://github.com/man-group/sparrow/blob/11ef0540b3227b79cf66fb0f331f493ff488b6de/include/sparrow/c_interface.hpp#L103C27-L103C48)

Also note that it would be best if that is a default parameter (as mentionned in another comment). That makes the ownership clear and irrelevant of how the objects are passed. I dont think it's a good idea in general to have 2 ways to specifies the same ownership, to be clear.

Reply:
Ok done

Reply:
removed finally: https://github.com/man-group/sparrow/pull/206#discussion_r1758423297